### PR TITLE
fix(paddle): freeze DPA3 descriptor parameters

### LINKED ISSUE #5686: [Code scan] Freeze Paddle DPA3 parameters with stop_gradient
This issue comes from a Codex global scan of `deepmodeling/deepmd-kit` at commit `73de44b1f94471b2e3bdb6b11f57b34d7bc791bb`.

## Problem

The Paddle DPA3 descriptor tries to honor `trainable=False` by setting `param.requires_grad = trainable`:

https://github.com/deepmodeling/deepmd-kit/blob/73de44b1f94471b2e3bdb6b11f57b34d7bc791bb/deepmd/pd/model/descriptor/dpa3.py#L251-L253

Other Paddle descriptors freeze parameters with `stop_gradient`, which is the Paddle parameter flag used by autograd:

https://github.com/deepmodeling/deepmd-kit/blob/73de44b1f94471b2e3bdb6b11f57b34d7bc791bb/deepmd/pd/model/descriptor/se_a.py#L528-L531

https://github.com/deepmodeling/deepmd-kit/blob/73de44b1f94471b2e3bdb6b11f57b34d7bc791bb/deepmd/pd/model/descriptor/dpa2.py#L331-L333

## Impact

Users configuring a non-trainable Paddle DPA3 descriptor can still update descriptor parameters during training because the freeze pass writes a PyTorch-style attribute instead of Paddle's `stop_gradient` flag. DPA3's RepFlow/MLP parameters rely on this outer descriptor freeze pass, so the requested freeze does not take effect.

## Suggested fix

Use the same Paddle pattern as the neighboring descriptors:

```python
for param in self.parameters():
    param.stop_gradient = not trainable
```

If Paddle exposes any additional trainable-state APIs in the supported versions, keep this consistent across all Paddle descriptor wrappers.

### PR DESCRIPTION
**Pull request by njzjz**
Closes #5686.

## Summary

- freeze Paddle DPA3 parameters with Paddle's `stop_gradient` flag instead of assigning a PyTorch-style `requires_grad` attribute;
- centralize descriptor-level trainability in `_apply_trainable()`;
- reapply the setting after deserialization replaces type, charge/spin, MLP, edge/angle, and RepFlow child layers;
- add a regression covering both direct construction and serialize/deserialize round-trip with charge/spin embeddings enabled.

## Root cause

Paddle's autograd contract is `parameter.stop_gradient`. On the project's minimum supported Paddle 3.0.0, assigning `parameter.requires_grad` merely adds an unrelated Python attribute and leaves `stop_gradient=False`, so optimizers can still update a descriptor configured with `trainable=False`.

Construction was not the only affected path. `DescrptDPA3.deserialize()` constructs a descriptor and then replaces several registered child layers with freshly deserialized parameters. The descriptor-level freeze therefore needs to run again after all replacements are complete.

## Why existing tests missed this

Existing DPA3 consistency tests compare numerical outputs and serialization behavior, but do not inspect Paddle autograd flags. A model can produce identical inference values while its supposedly frozen parameters remain trainable.

The regression temporarily masks newer Paddle's `requires_grad` compatibility alias, retaining the minimum-version semantics even in nightly CI. It then requires every named parameter to have `stop_gradient=True` both immediately after construction and after a round-trip. Enabling `add_chg_spin_ebd=True` ensures the optional replacement paths are included.

## Old-fail / new-pass validation

- original implementation, Paddle 3.0.0: failed with 17 parameters still requiring gradients;
- original implementation, Paddle `3.4.0.dev20260310`: failed with the same 17 parameters under the minimum-version compatibility seam;
- constructor-only intermediate fix, Paddle 3.0.0: construction passed but round-trip failed with 22 unfrozen parameters;
- constructor-only intermediate fix, Paddle `3.4.0.dev20260310`: round-trip failed with 28 unfrozen parameters;
- complete fix, Paddle 3.0.0: regression passed, including both subtests;
- complete fix, Paddle `3.4.0.dev20260310`: regression plus existing DPA3 consistency test passed (`2 passed`).

Additional checks:

- `ruff format .`;
- `ruff check .`;
- `git diff --check`;
- independent final review: no findings.

Coding agent: Codex
Codex version: codex-cli 0.144.1
Model: gpt-5.6-sol
Reasoning effort: xhigh

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed non-trainable DPA3 descriptor parameters becoming trainable after deserialization or type-map updates.
  * Ensured non-trainable settings consistently prevent gradient updates during initialization, model loading, and descriptor updates.
  * Preserved non-trainable behavior for charge, spin, and mixing embeddings.
  * Prevented parameter sharing between descriptors with incompatible trainability settings.

* **Tests**
  * Added coverage for parameter gradient settings during initialization, serialization/deserialization, type-map changes, and parameter sharing.

### CODE REVIEW

**Comment by njzjz** [comment]
Possible reviewers based on changed lines, exact file history, and exact-file review history:

- @iProzd — 8 reviews on exact changed files (`deepmd/pd/model/descriptor/dpa3.py`, `source/tests/pd/model/test_dpa3.py`).

No review request was made automatically.

Coding agent: Codex
Codex version: codex-cli 0.144.4
Model: gpt-5.6-sol
Reasoning effort: xhigh

File: deepmd/pd/model/descriptor/dpa3.py [hunk 1]
```diff
@@ -255,11 +255,14 @@ def init_subclass_params(sub_data: dict | Any, sub_class: type) -> Any:
...
-        # set trainable
-        for param in self.parameters():
-            param.requires_grad = trainable
+        self._apply_trainable()
         self.compress = False
 
+    def _apply_trainable(self) -> None:
+        """Apply the descriptor-level trainable setting to every parameter."""
+        for param in self.parameters():
+            param.stop_gradient = not self.trainable
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pd/model/descriptor/dpa3.py:267]: Fixed in 7109a5453. `DescrptDPA3.change_type_map()` now reapplies the descriptor-level trainable setting immediately after the type embedding replaces its first matrix. A regression remaps a frozen descriptor and verifies that no parameter becomes optimizer-visible. Targeted Ruff and `git diff --check` pass; the Paddle tests could not run locally because Paddle is not installed in this environment.

Coding agent: Codex
Codex version: codex-cli 0.144.6
Model: gpt-5.6-sol
Reasoning effort: xhigh

File: deepmd/pd/model/descriptor/dpa3.py [hunk 2]
```diff
@@ -557,6 +560,9 @@ def t_cvt(xx: Any) -> paddle.Tensor:
         obj.repflows.layers = paddle.nn.LayerList(
             [RepFlowLayer.deserialize(layer) for layer in repflow_layers]
         )
+        # Deserialization replaces several sublayers after construction, so apply
+        # the descriptor-level setting again to their newly registered parameters.
+        obj._apply_trainable()
```
[the remark this answers is not in the document]
Comment (njzjz) [reply deepmd/pd/model/descriptor/dpa3.py:576]: Fixed in 7109a5453. The optional charge embedding, spin embedding, and charge/spin mixing MLP now receive the outer `trainable` setting at construction time, so their serialized child dictionaries preserve the effective frozen state for cross-backend loads. The regression checks those serialized flags as well as constructed and deserialized Paddle parameters. Targeted Ruff and `git diff --check` pass; the Paddle tests could not run locally because Paddle is not installed in this environment.

Coding agent: Codex
Codex version: codex-cli 0.144.6
Model: gpt-5.6-sol
Reasoning effort: xhigh
